processing/src/main/java/io/druid/query/groupby/epinephelinae/GroupByRowProcessor.java

```diff
@@ -0,0 +1,508 @@
+                      }
+                  );
+                  filteredSequence.accumulate(grouper, accumulator);
+
+                  return new CloseableGrouperIterator<>(
+                      grouper,
+                      true,
+                      new Function<Grouper.Entry<GroupByMergingKey>, Row>()
+                      {
+                        @Override
+                        public Row apply(Grouper.Entry<GroupByMergingKey> entry)
+                        {
+                          Map<String, Object> theMap = Maps.newLinkedHashMap();
+
+                          // Add dimensions.
+                          for (int i = 0; i < entry.getKey().getDimensions().length; i++) {
+                            theMap.put(
+                                query.getDimensions().get(i).getOutputName(),
+                                Strings.emptyToNull(entry.getKey().getDimensions()[i])
+                            );
+                          }
+
+                          // Add aggregations.
+                          for (int i = 0; i < entry.getValues().length; i++) {
+                            theMap.put(query.getAggregatorSpecs().get(i).getName(), entry.getValues()[i]);
+                          }
+
+                          return new MapBasedRow(
+                              query.getGranularity().toDateTime(entry.getKey().getTimestamp()),
+                              theMap
+                          );
+                        }
+                      }
+                  );
+                }
+
+                @Override
+                public void cleanup(CloseableGrouperIterator<GroupByMergingKey, Row> iterFromMake)
+                {
+                  iterFromMake.close();
+                }
+              }
+          ),
+          new Closeable()
+          {
+            @Override
+            public void close() throws IOException
+            {
+              mergeBufferHolder.close();
+              CloseQuietly.close(temporaryStorage);
+            }
+          }
+      );
+    }
+    catch (Exception e) {
+      // Exception caught while creating the sequence; release resources.
+      mergeBufferHolder.close();
+      CloseQuietly.close(temporaryStorage);
+      throw e;
+    }
+  }
+
+  private static class RowBasedColumnSelectorFactory implements ColumnSelectorFactory
+  {
+    private Row row = null;
+
+    public void setRow(Row row)
+    {
+      this.row = row;
+    }
+
+    @Override
+    public DimensionSelector makeDimensionSelector(
+        DimensionSpec dimensionSpec
+    )
+    {
+      return dimensionSpec.decorate(makeDimensionSelectorUndecorated(dimensionSpec));
+    }
+
+    private DimensionSelector makeDimensionSelectorUndecorated(
+        DimensionSpec dimensionSpec
+    )
+    {
+      final String dimension = dimensionSpec.getDimension();
+      final ExtractionFn extractionFn = dimensionSpec.getExtractionFn();
+
+      return new DimensionSelector()
```

Comment:
Added a failing test for outer filter aggregator, I'll open an issue for it after this PR gets merged.